pkg/controller/etcdproxy/certs.go

```diff
@@ -0,0 +1,252 @@
+package etcdproxy
+
+import (
+	"crypto/x509/pkix"
+	"fmt"
+	"math/rand"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	etcdstoragev1alpha1 "github.com/xmudrii/etcdproxy-controller/pkg/apis/etcd/v1alpha1"
+	"github.com/xmudrii/etcdproxy-controller/pkg/certs"
+)
+
+// newEtcdProxyCertificates generates new self-signed server signer and server certificate/key pair.
+// Self-signed server signer cert is stored in APIServer ConfigMap to be used as CA by etcd, while server cert/key pair
+// is stored in Controller Secret to be used by etcdproxy.
+func (c *EtcdProxyController) newEtcdProxyCertificates(etcdstorage *etcdstoragev1alpha1.EtcdStorage) error {
+	currentTime := time.Now
+	r := rand.New(rand.NewSource(currentTime().UnixNano()))
+	serviceUrl := fmt.Sprintf("%s.%s.svc", serviceName(etcdstorage), c.config.ControllerNamespace)
+
+	// TODO: implement certificates regeneration. this could require architecture change, because we could require
+	// new secret for storing singer key.
+	serverSigner, err := certs.NewCACertificate(pkix.Name{
+		CommonName: fmt.Sprintf("%s-server-signer-%v", serviceUrl, time.Now().Unix()),
+	}, r.Int63n(100000), currentTime)
+	if err != nil {
+		return err
+	}
+	serverSignerCert, err := certs.EncodeCertificates(serverSigner.Certificates...)
+	if err != nil {
+		return err
+	}
+
+	// Generate server certificate/key pair.
+	serverBundle, err := serverSigner.NewServerCertificate(pkix.Name{CommonName: serviceUrl},
+		[]string{serviceUrl}, r.Int63n(100000), currentTime)
+	if err != nil {
+		return err
+	}
+	serverCertBytes, err := certs.EncodeCertificates(serverBundle.Certificates...)
+	if err != nil {
+		return err
+	}
+	serverKeyBytes, err := certs.EncodeKey(serverBundle.Key)
+	if err != nil {
+		return err
+	}
+
+	// Write pairs to appropriate ConfigMaps and Secrets.
+	err = c.updateAPIServerServingCAConfigMaps(etcdstorage, serverSignerCert)
```

Comment:
site effects in a constructor like func `new....` is not good.

Reply:
Maybe rename it to `setNew...`

Reply:
The function name are updated to `setNew...`.